src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs

```diff
@@ -101,43 +109,93 @@ internal CloudTable GetDiagnosticEventsTable(DateTime? now = null)
             return _diagnosticEventsTable;
         }
 
-        private static string GetCurrentTableName(DateTime now)
+        private static string GetTableName(DateTime date)
         {
-            return $"{TableNamePrefix}{now:yyyyMM}";
+            return $"{TableNamePrefix}{date:yyyyMM}";
         }
 
         protected internal virtual async void OnFlushLogs(object state)
         {
             await FlushLogs();
         }
 
+        private async Task PurgePreviousEventVersions()
+        {
+            _logger.LogTrace("Purging previous diagnostic event versions.");
+
+            try
+            {
+                var tables = await TableStorageHelpers.ListTablesAsync(TableClient, TableNamePrefix);
+
+                foreach (var table in tables)
+                {
+                    var tableRecords = await table.ExecuteQuerySegmentedAsync(new TableQuery<DiagnosticEvent>(), null);
+
+                    // Skip tables that have 0 records
+                    if (tableRecords.Results.Count == 0)
+                    {
+                        continue;
+                    }
+
+                    // Delete table if it doesn't have records with EventVersion
+                    var eventVersionDoesNotExists = tableRecords.Results.Any(record => string.IsNullOrEmpty(record.EventVersion) == true);
+                    if (eventVersionDoesNotExists)
+                    {
+                        _logger.LogTrace("Deleting table '{tableName}' as it contains records without an EventVersion.", table.Name);
+                        await table.DeleteIfExistsAsync();
+                        continue;
+                    }
+
+                    // If the table does have EventVersion, query if it is an outdated version
+                    var eventVersionOutdated = tableRecords.Results.Any(record => record.EventVersion.CompareTo(DiagnosticEvent.CurrentEventVersion) < 0);
+                    if (eventVersionOutdated)
+                    {
+                        _logger.LogTrace("Deleting table '{tableName}' as it contains records with an outdated EventVersion.", table.Name);
+                        await table.DeleteIfExistsAsync();
+                    }
+                }
+
+                _purged = true;
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error occurred when attempting to purge previous diagnostic event versions.");
```

Comment:
Should we also log, how many tables were there and how many were successfully purged? Wondering that would be helpful in the case of partial purge and and an exception occurred.

We can do that logging in a finally block.

Reply:
Not every table we find might need to be deleted so we would have to track something like `identifiedForDeletionCount` and `deletedTableCount` and report those at the end.

I don't think we need this log since we already have logs before each table gets deleted saying "Deleting table {blah}" And I think from those logs it's enough to identify if a table that was supposed to be deleted was or wasn't. I think that is sufficient, what do you think?